### PR TITLE
Add async ssa graph executor

### PR DESCRIPTION
**Pull request by jacquesqiao**
image

### CODE REVIEW

File: paddle/fluid/framework/parallel_executor.cc [hunk 1]
```diff
@@ -401,14 +419,22 @@ void ParallelExecutor::BCastParamsToDevices(
...
-        // FIXME(zcd): LR_DECAY_COUNTER should not be shared. This is a hot fix.
-        if (member_->use_all_reduce_ || member_->use_cuda_ ||
-            var == "@LR_DECAY_COUNTER@") {
+        auto share_memory = [&] {
           t->Resize(dims);
           t->mutable_data(cpu, main_tensor.type());
           paddle::framework::TensorCopy(main_tensor, cpu, t);
+        };
+
+        auto copy_memory = [&] { t->ShareDataWith(main_tensor); };
```
Comment (typhoonzero) [inline paddle/fluid/framework/parallel_executor.cc]: seems `copy_memory` and `share_memory` are reversed?

Comment (jacquesqiao) [reply paddle/fluid/framework/parallel_executor.cc]: fix done

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+// limitations under the License.
+
+#include "paddle/fluid/framework/details/async_ssa_graph_executor.h"
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+AsyncSSAGraphExecutor::AsyncSSAGraphExecutor(
+    const ExecutionStrategy &strategy, const std::vector<Scope *> &local_scopes,
```
Comment (guru4elephant) [inline paddle/fluid/framework/details/async_ssa_graph_executor.cc:43]: local_scopes是从哪里创建带入的？

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  for (size_t fetch_idx = 0; fetch_idx < fetch_tensors.size(); ++fetch_idx) {
+    std::vector<const LoDTensor *> lodtensor_ptrs;
+    lodtensor_ptrs.reserve(local_scopes_.size());
+    for (size_t scope_idx = 0; scope_idx < local_scopes_.size(); ++scope_idx) {
+      lodtensor_ptrs.push_back(&fetch_data.at(scope_idx).at(fetch_idx));
+    }
+    ret.emplace_back();
+    ret.back().MergeLoDTensor(lodtensor_ptrs, platform::CPUPlace());
```
Comment (guru4elephant) [inline paddle/fluid/framework/details/async_ssa_graph_executor.cc:139]: num_iteration_per_run_ > 1的情况下，各线程执行速度不一致，merge各个local_scope的结果是否有意义？

Comment (jacquesqiao) [reply paddle/fluid/framework/details/async_ssa_graph_executor.cc:139]: 这个感觉可以去掉其实，反正已经是纯异步了，相当于减少一点做evel的数据量

Comment (jacquesqiao) [reply paddle/fluid/framework/details/async_ssa_graph_executor.cc:139]: 大家步调不一致，参数版本也不一致，确实应该去掉，观察其中一个线程就够了

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/details/async_ssa_graph_executor.cc:43]: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/parallel_executor.cc#L217

File: paddle/fluid/framework/parallel_executor.cc [hunk 4]
```diff
@@ -258,81 +264,118 @@ ParallelExecutor::ParallelExecutor(
...
   // ncclOp
-  std::unique_ptr<ir::Graph> graph;
+  std::vector<std::unique_ptr<ir::Graph>> graphs;
 #if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
-  graph = build_strategy.Apply(main_program, member_->places_, loss_var_name,
-                               member_->local_scopes_, member_->nranks_,
-                               member_->use_cuda_, member_->nccl_ctxs_.get());
+  if (build_strategy.async_mode_ && !build_strategy.is_distribution_) {
+    VLOG(3) << "use local async mode";
+    for (size_t i = 0; i < member_->places_.size(); ++i) {
```
Comment (Yancey0623) [inline paddle/fluid/framework/parallel_executor.cc]: @panyx0718 has a PR that passed graph insetead of program: https://github.com/PaddlePaddle/Paddle/pull/15425 . And ParallelGraphExecutor does not dependence program_desc: https://github.com/PaddlePaddle/Paddle/pull/15716 ?

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      run_futures.emplace_back(pool_->enqueue(std::move(call)));
+    } else {
+      fetch_data.emplace_back(std::move(call()));
+    }
+  }
+
+  if (pool_) {
+    for (auto &f : run_futures) {
+      if (exception_holder_.IsCaught()) {
+        f.wait();
```
Comment (panyx0718) [inline paddle/fluid/framework/details/async_ssa_graph_executor.cc]: why sync here?

File: paddle/fluid/framework/details/build_strategy.h [hunk 6]
```diff
@@ -90,6 +90,7 @@ struct BuildStrategy {
   // num_trainers is 1, so the current fields of build_strategy doesn't tell if
   // it's distributed model.
   bool is_distribution_{false};
+  bool async_mode_{false};
```
Comment (panyx0718) [inline paddle/fluid/framework/details/build_strategy.h:94]: what is the relationship between async_mode and is_distribution

File: paddle/fluid/framework/details/build_strategy.h [hunk 7]
```diff
@@ -90,6 +90,7 @@ struct BuildStrategy {
...
   int num_trainers_{1};
```
Comment (panyx0718) [inline paddle/fluid/framework/details/build_strategy.h:95]: can num_trainers > 1 and not is_distribution?

File: paddle/fluid/framework/parallel_executor.cc [hunk 8]
```diff
@@ -187,7 +188,7 @@ ParallelExecutor::ParallelExecutor(
     const std::string &loss_var_name, Scope *scope,
     const std::vector<Scope *> &local_scopes,
     const ExecutionStrategy &exec_strategy, const BuildStrategy &build_strategy,
-    ir::Graph *graph)
+    std::vector<ir::Graph *> graphs)
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc]: avoid multiple graphs. A single graph can contain multiple sub-graphs

File: paddle/fluid/framework/parallel_executor.cc [hunk 9]
```diff
@@ -257,20 +264,50 @@ ParallelExecutor::ParallelExecutor(
...
 #if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
-
-  temp_owned_graph = build_strategy.Apply(
-      std::move(temp_owned_graph), member_->places_, loss_var_name,
-      member_->local_scopes_, member_->nranks_, member_->use_cuda_,
-      member_->nccl_ctxs_.get());
+  if (build_strategy.async_mode_ && !build_strategy.is_distribution_) {
+    VLOG(3) << "use local async mode";
+    temp_owned_graph =
+        build_strategy.Apply(std::move(temp_owned_graph), {member_->places_[0]},
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc:283]: why each graph needs to go through multi-device pass?

File: python/paddle/fluid/parallel_executor.py [hunk 10]
```diff
@@ -177,12 +177,17 @@ def place_obj(place):
 
         # step7: init ParallelExecutor
         # ParallelExecutor API will be deprecated, don't support parallel graph.
-        self._graph = core.Graph(main.desc)
+        self._graphs = []
```
Comment (panyx0718) [inline python/paddle/fluid/parallel_executor.py]: parallel_executor.py is deprecated.

File: paddle/fluid/framework/parallel_executor.h [hunk 11]
```diff
@@ -50,7 +50,7 @@ class ParallelExecutor {
                             const std::vector<Scope *> &local_scopes,
                             const ExecutionStrategy &exec_strategy,
                             const BuildStrategy &build_strategy,
-                            ir::Graph *graph);
+                            std::vector<ir::Graph *> graphs);
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.h]: don't do multiple graphs.

**Review by panyx0718 — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/reader/buffered_reader.cc [hunk 12]
```diff
@@ -20,6 +20,7 @@ namespace paddle {
 namespace operators {
 namespace reader {
 BufferedReader::~BufferedReader() {
+  VLOG(1) << "~BufferedReader";
```
Comment (seiriosPlus) [inline paddle/fluid/operators/reader/buffered_reader.cc:24]: delete this ```VLOG```

File: python/paddle/fluid/parallel_executor.py (hunk 10 above)
Comment (seiriosPlus) [inline python/paddle/fluid/parallel_executor.py]: the comment is wrong.

**Pull request merged**